### PR TITLE
Introduce tiny types

### PR DESCRIPTION
**Pull request by uglybugger**
This PR is to introduce the concept of tiny types to this library.

There are some upcoming changes to Octopus Server that will be helped tremendously by the ability to rely on strongly-typed reference collections and these changes are in support of that effort.

There is a base `TinyType<T>` implementation (with tests around equality comparisons etc.) and three concrete usages: `WorkerPoolId`, `WorkerPoolName` and `WorkerPoolIdOrName`. The first two should be self-explanatory; the last is a not-quite-discriminated-union because we need to support references to worker pools either via name or ID and we won't know which form it takes until resolved at deployment time.

### CODE REVIEW

**Comment by uglybugger** [comment]
This PR is currently blocked by #33, hence the currently-broken branch build.

File: source/Node.Extensibility.Tests/TinyTypes/StringEqualityTests.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
+﻿using System.Collections.Generic;
+using NUnit.Framework;
+using Octopus.Server.Extensibility.TinyTypes;
+
+namespace Node.Extensibility.Tests.TinyTypes
+{
+    public class StringEqualityTests
```
Comment (matt-richardson) [inline source/Node.Extensibility.Tests/TinyTypes/StringEqualityTests.cs:7]: Should these be in here? Or should they be in `TinyTypes` itself?

File: source/Server.Extensibility/HostServices/Model/WorkerPools/WorkerPoolId.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+namespace Octopus.Server.Extensibility.HostServices.Model.WorkerPools
+{
+    public class WorkerPoolId : CaseInsensitiveStringTinyType
+    {
+        public WorkerPoolId(string value) : base(value)
+        {
+            if (string.IsNullOrWhiteSpace(value))
+                throw new ArgumentException(nameof(value),
+                    "Worker pool IDs or Names should never be null or empty. If the value you want to communicate is null/empty/whitespace then " +
+                    "just pass a null rather than a tiny type wrapping it.");
```
Comment (matt-richardson) [inline source/Server.Extensibility/HostServices/Model/WorkerPools/WorkerPoolId.cs:13]: Can we recommend they create a `NullWorkerPoolId` instance instead? (ie NullObject pattern)... 
I'd hate to see us take 2 steps forwards (with tinytypes) but then double down on NullRefExceptions

File: source/Server.Extensibility/Server.Extensibility.csproj [hunk 3]
```diff
@@ -11,6 +11,7 @@
     <PackageIconUrl>http://i.octopusdeploy.com/resources/Avatar3_360.png</PackageIconUrl>
     <PackageLicenseUrl>https://github.com/OctopusDeploy/ServerExtensibility/blob/master/LICENSE.txt</PackageLicenseUrl>
     <PackageProjectUrl>https://github.com/OctopusDeploy/ServerExtensibility</PackageProjectUrl>
+    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
```
Comment (matt-richardson) [inline source/Server.Extensibility/Server.Extensibility.csproj:14]: Unclear why this is included in this PR?

File: source/Server.Extensibility/TinyTypes/TinyType.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    ///     Provides a strongly-typed wrapper around a (ideally value) type.
+    /// </summary>
+    /// <remarks>
+    ///     NOTE: We deliberately do not provide implicit or explicit cast operators. Implicit cases would defeat the purpose
+    ///     of providing compile-time errors when disparate types are compared or passed as parameters. Explicit cast operators
+    ///     would not cause this problem... but they would be *so very tempting* that it's inevitable that they'd end up being
+    ///     used to coerce types between values that really shouldn't be coerced. The design decision at present is that tiny
+    ///     types must be explicitly unwrapped and re-wrapped in order to minimise the likelihood of bugs' being created.
+    /// </remarks>
+    public abstract class TinyType<T>
```
Comment (matt-richardson) [inline source/Server.Extensibility/TinyTypes/TinyType.cs:15]: 'ang on, isn't TinyTypes a library? Why is it inlined here?

File: source/Node.Extensibility.Tests/TestResults/slewi_SHANNONXPS_2017-10-30_11_57_13.trx [hunk 5]
```diff
@@ -1,68 +0,0 @@
-﻿<?xml version="1.0" encoding="UTF-8"?>
```
Comment (matt-richardson) [inline source/Node.Extensibility.Tests/TestResults/slewi_SHANNONXPS_2017-10-30_11_57_13.trx:1]: Not sure what this change has to do with this PR?

File: source/Node.Extensibility.Tests/TinyTypes/StringEqualityTests.cs (hunk 1 above)
Comment (uglybugger) [reply source/Node.Extensibility.Tests/TinyTypes/StringEqualityTests.cs:7]: Good point. I did think about this :)

I'd expect that we'd pull these out into the TinyTypes library soon-ish but there are some design differences (mainly around implicit cast operators) that I'd like to discuss first. I didn't want to introduce yet another dependency in the chain that would be so far removed from where we're actually trying to use it, which is already one consumer upstream from here.

File: source/Server.Extensibility/TinyTypes/TinyType.cs (hunk 4 above)
Comment (uglybugger) [reply source/Server.Extensibility/TinyTypes/TinyType.cs:15]: Oops - I should have answered this question first. Copy/paste from my comment above:

I'd expect that we'd pull these out into the TinyTypes library soon-ish but there are some design differences (mainly around implicit cast operators) that I'd like to discuss first. I didn't want to introduce yet another dependency in the chain that would be so far removed from where we're actually trying to use it, which is already one consumer upstream from here.

File: source/Node.Extensibility.Tests/TestResults/slewi_SHANNONXPS_2017-10-30_11_57_13.trx (hunk 5 above)
Comment (uglybugger) [reply source/Node.Extensibility.Tests/TestResults/slewi_SHANNONXPS_2017-10-30_11_57_13.trx:1]: It was an extraneous test results file that looks like it was committed by accident a while ago. I can separate that out into a separate PR if it offends enough people by leaving it in here :)

File: source/Server.Extensibility/HostServices/Model/WorkerPools/WorkerPoolId.cs (hunk 2 above)
Comment (uglybugger) [reply source/Server.Extensibility/HostServices/Model/WorkerPools/WorkerPoolId.cs:13]: I think that one's a beer o'clock discussion. Happy to have it, though :)

**Comment by uglybugger** [comment]
Spoke with @slewis74 via Zoom. Verbal approval - happy that it's moving us in the right direction. We're good to merge.

**Pull request merged**